src/plugins/remark-code-import.ts

```diff
@@ -81,55 +83,90 @@ function include(includePath: string, anchor?: string): string {
   try {
     let fileContent = fs.readFileSync(includePath, "utf8");
 
-    // if the anchor is in the format "start:end", extract the lines between the start and end
-    if (anchor && anchor.search(":") != -1) {
-      const [start, end] = anchor.split(":");
-      const lines = fileContent.split("\n");
-      const startIndex = start ? parseInt(start) - 1 : 0;
-      const endIndex = end ? parseInt(end) : lines.length;
-      fileContent = lines.slice(startIndex, endIndex).join("\n");
-    } else if (anchor) {
-      // If an anchor is specified, extract the relevant section from the file between
-      // the start and end anchor comments. This is done using a
-      const commentRegex = rx`
+    if (anchor) {
+      if (anchor.search(":") != -1) {
+        // if the anchor is in the format "start:end", extract the lines between the start and end
+        const [start, end] = anchor.split(":");
+        fileContent = extractLines(fileContent, start, end);
+      } else if (anchor.endsWith("()")) {
+        // If an anchor is specified with a function call, extract the relevant section from the file between
+        // the start of the function and the end of the function.
+        const name = anchor.slice(0, -2);
+        fileContent = extractFunction(fileContent, name);
+      } else {
+        // If an anchor is specified, extract the relevant section from the file between
+        // the start and end anchor comments. This is done using a
+        fileContent = extractBetweenComments(fileContent, anchor);
+      }
+    }
+    // Remove lines containing start and end anchor comments
+    return fileContent
+      .split("\n")
+      .filter((line) => !line.includes("// ANCHOR: ") && !line.includes("// ANCHOR_END: "))
+      .join("\n");
+  } catch (err) {
+    let message = (err as Error).message;
+    throw new Error(`Unable to include file '${includePath}'. ${message}`);
+  }
+}
+
+function extractLines(content: string, start: string, end: string) {
+  const lines = content.split("\n");
+  const startIndex = start ? parseInt(start) - 1 : 0;
+  const endIndex = end ? parseInt(end) : lines.length;
+  content = lines.slice(startIndex, endIndex).join("\n");
+  return content;
+}
+
+function extractFunction(content: string, name: string) {
+  const functionRegex = new RegExp(`\\s*(fn)\\s+${name}\\s*\\(.*?\\).*?{`, "s");
+  const match = functionRegex.exec(content);
+  if (match) {
+    const startIndex = match.index + match[0].length + 1;
+    let braceCount = 1;
+    let endIndex = startIndex;
+    for (let i = startIndex; i < content.length; i++) {
+      if (content[i] === "{") braceCount++;
+      if (content[i] === "}") braceCount--;
+      if (braceCount === 0) {
+        endIndex = i + 1;
+        break;
+      }
+    }
+    return content.substring(match.index, endIndex);
+  }
+  throw new Error(`Function '${name}' not found`);
+}
```

Comment:
Maybe skip counting braces that appear in comments, strings, or character literals? It'll make it more robust to code like this:

```rust
fn main() {
  println!(" this is an opening brace: { ")
}
```

Reply:
yeah - there's certainly lots to make this robust. I'd not worry too much about that specific problem at least at first, as unbalanced braces should come up rarely enough to be able to just deal with them by fixing the code. The other parts of the TODO are much more necessary.

Reply:
Using tree-sitter fixes this properly.
[1ce9d6b](https://github.com/ratatui/ratatui-website/pull/708/commits/1ce9d6b0bb485500ae6898968b65515713850335)